### DISCUSSION TITLE
Unexpected error thrown in Next.js core during `findHeadInCache` call

### DISCUSSION THREAD

**Discussion by l-you** [vercel/next.js#54180] [App Router] [answered]
Function `findHeadInCache` throws an error during navigation back. 
Look at following part of Next.js core. https://github.com/vercel/next.js/blob/df1de76c5cbb38c9eac9ddec59bfbc9d8559d944/packages/next/src/client/components/app-router.tsx#L456
**Why there is no validation to check whether `tree[1]` is undefined?**
Error message says `TypeError: undefined is not an object (evaluating 'Object.keys(n)')` where `n` in my error message is actually `tree[1]`.
Issue happens during navigation back.

Please look at following minified production code. Error source marked with comment "//ERROR THROWN HERE".
```
   Object.defineProperty(t, "__esModule", {
            value: !0
        });
        var r = n(84337);
        Object.defineProperty(t, "__esModule", {
            value: !0
        }),
        Object.defineProperty(t, "findHeadInCache", {
            enumerable: !0,
            get: function() {
                return function e(t, n) {
                    if (0 === Object.keys(n).length) //ERROR THROWN HERE
                        return t.head;
                    for (var o in n) {
                        var i = r._(n[o], 2),
                            u = i[0],
                            s = i[1],
                            c = t.parallelRoutes.get(o);
                        if (c) {
                            var l = (0, a.createRouterCacheKey)(u),
                                f = c.get(l);
                            if (f) {
                                var d = e(f, s);
                                if (d)
                                    return d
                            }
                        }
                    }
                }
            }
        });
        var a = n(28579);
        ("function" == typeof t.default || "object" == typeof t.default && null !== t.default) && void 0 === t.default.__esModule && (Object.defineProperty(t.default, "__esModule", {
            value: !0
        }), Object.assign(t.default, t), e.exports = t.default)
    },
    97602: function(e, t) {
        "use strict";
        function n(e) {
            return Array.isArray(e) ? e[1] : e
        }
        Object.defineProperty(t, "__esModule", {
            value: !0
        }),
        Object.defineProperty(t, "getSegmentValue", {
            enumerable: !0,
            get: function() {
                return n
            }
        }),
```

**Comment by icyJoseph** [comment]
Hi,

How do you trigger this? Just navigating back? Could you give some series of steps?

I mean sure, that should be handled, but maybe it is __expected to not ever be undefined__? like we'd have broken functionality anyway, even if handled

**Reply by l-you** [reply]
Please take a look at https://ladyboom.ua .

1. Go to main page
2. Click first link
3. Navigate back with buttons.

Issue can be reproduced on both MacOS Safari and IOS Safari.
Here is screen recording with a proof. We Use Next.js 13.4.17.
Error is not reproducible on dev server.

https://github.com/vercel/next.js/assets/32598874/2e67b5d5-2099-4fb2-8c53-99e6ed6e28c7

**Reply by icyJoseph** [reply]
🤔 I just deployed to two apps using 13.4.17 and none has this issue. 

I also reproduced yours in Chrome.

**Reply by icyJoseph** [reply]
And now it works in Chrome, but fails in Safari... wtf mmm

**Reply by icyJoseph** [reply]
I see that navigating to the first link is not an SPA navigation. Could this be the issue?

![Screenshot 2023-08-17 at 18 08 53](https://github.com/vercel/next.js/assets/21013447/5187552f-77ba-4e47-861b-5dae2027c721)

Above a screenshot of Chrome failing, it is random in Chrome.

**Reply by icyJoseph** [reply]
And now I got one of my deploys to fail...

**Reply by icyJoseph** [reply]
To reproduce this locally:

- run `npm run build && npm run start`, 
- when it runs open a new tab in your browser,
- type localhost:3000
- App loads normally
- Browse back with the back button
- Browse forward with the forward button
- App crashes

**Reply by l-you** [reply]
@icyJoseph Can we roll back our production to the previous version of Next.js? We hope reproduction you provided works in 100% of cases.
> To reproduce this locally:
> 
> * run `npm run build && npm run start`,
> * when it runs open a new tab in your browser,
> * type localhost:3000
> * App loads normally
> * Browse back with the back button
> * Browse forward with the forward button
> * App crashes

**Reply by icyJoseph** [reply]
Assuming it is the same issue, then yes I think so. I tried with next@13.4.16 and could not reproduce the same I described.

**Reply by icyJoseph** [reply]
Please report back if rolling back works. I am trying to find the commit that might have introduced this.

**Reply by icyJoseph** [reply, marked as answer]
Opened an issue here: https://github.com/vercel/next.js/issues/54184

**Reply by l-you** [reply]
@icyJoseph We rolled back our production. **Issue is not reproducible with next@13.4.16**
